# Data Prepocessing

In [2]:
import numpy as np
import pandas as pd
import csv
from sklearn import preprocessing
from sklearn.decomposition import PCA

## Bank Dataset

In [197]:
# Read data from file

with open('bank/bank-full.csv', 'r') as csvfile:
    spamreader = csv.reader(csvfile, delimiter=';')
    df = list(spamreader)
    df = np.asarray(df)
    df = pd.DataFrame(data=df[1:], columns=df[0])
    

In [198]:
# Inspecting the unknown values in the dataset

jobs = df['job']
unknown_jobs = 0
for job in jobs:
    if job == 'unknown':
        unknown_jobs = unknown_jobs + 1
        
educations = df['education']
unknown_education = 0
for education in educations:
    if education == 'unknown':
        unknown_education = unknown_education + 1

defaults = df['default']
unknown_default = 0
for default in defaults:
    if default == 'unknown':
        unknown_default = unknown_default + 1

contacts = df['contact']
unknown_contact = 0
for contact in contacts:
    if contact == 'unknown':
        unknown_contact = unknown_contact + 1
    
pdays = df['pdays']
unknown_pday = 0
for pday in pdays:
    if pday == '-1':      # Treating never contacted as unknown
        unknown_pday = unknown_pday + 1

pouts = df['poutcome']
unknown_pouts = 0
for pout in pouts:
    if pout == 'unknown':
        unknown_pouts = unknown_pouts + 1
    

print("Unknown values\n")
print("Unknown jobs :", unknown_jobs)
print("Unknown education :", unknown_education)
print("Unknown contacts :", unknown_contact)
print("Unknown pdays :", unknown_pday)
print("Unknown pouts :", unknown_pouts)



Unknown values

('Unknown jobs :', 288)
('Unknown education :', 1857)
('Unknown contacts :', 13020)
('Unknown pdays :', 36954)
('Unknown pouts :', 36959)


### Observations from Data Set

- Total number of rows: 45211
- Total number of attributes: 16 + output attribute (y)
- Yes samples - 5289
- No samples - 39922
- The number of yes samples are much smaller than the number of no samples indicating that the data set is highly skewed.

- Unknown values :-
    - job - 288
    - education - 1857
    - contact - 13020
    - pdays - 36954
    - poutcome - 36959
- A very large number of rows have pdays = -1 and poutcome = unknown. These represent clients that have not been contacted in the past

In [199]:
df = df.drop('duration', axis=1)             # Removing duration attribute

# Separating columns as numerical and categorical
numeric = ['age', 'balance', 'campaign', 'pdays', 'previous']
categorical = ['job', 'marital', 'education', 'default', 'housing', 'loan', 'contact', 'day', 'month', 'poutcome', 'y']


In [200]:
# Substituting pdays = 999 for pdays = -1
pdays = df['pdays']
pdays = pdays.replace('-1', '999')
df['pdays'] = pdays


# Substituting pdays = 999 and poutcome = failure for pdays = -1 and poutcome = unknown
# prev_list = ['pdays', 'poutcome']
# prev_df = df[prev_list]
# prev_df = prev_df.replace({'pdays':'-1', 'poutcome':'unknown'}, {'pdays':'999', 'poutcome':'failure'})
# df[prev_list] = prev_df


In [201]:
# Taking care of categorical values

# factorize assigns unique numeric values to each category
for c in categorical:
    df[c] = pd.factorize(df[c])[0]
    
# Use dummy attributes for categorical values
# df = pd.get_dummies(df, columns=categorical, prefix=categorical)
# df['y'] = pd.factorize(df['y'])

df[numeric] = df[numeric].astype('float')
df[categorical] = df[categorical].astype('category')

# df['y'] = df['y'].astype(int) # Converting otuput variable to integers


In [202]:
# Normalizing numerical values

scaler = preprocessing.StandardScaler()
rescaled_df = scaler.fit_transform(df[numeric])
df[numeric] = pd.DataFrame(rescaled_df, columns=numeric)

In [203]:
# Principal Component Analysis

y_values = df.loc[:, ['y']].values
df.drop(columns=['y'], axis=1, inplace=True)

pca = PCA()
#pca = PCA(n_components=10)
components = pca.fit_transform(df)
var_ratios = pca.explained_variance_ratio_
#rint(var_ratios)

# Calculating the number of attributes for 80% variance
s=0
count = 0
for ratio in var_ratios:
    s = s + ratio
    count = count + 1
    if s >= 0.8:
        break
print(count)

# Calculating the number of attributes for 95% variance
s=0
count1 = 0
for ratio in var_ratios:
    s = s + ratio
    count1 = count1+ 1
    if s >= 0.95:
        break
print(count1)

#df = pd.DataFrame(data=components)
df['y'] = y_values

2
5


### Observations from PCA

- There were 15 attributes (after removing duration) subjected to PCA
- 80% of the variance can be captured by 2 attributes after PCA
- 95% of the variance can be captured by 5 attributes after PCA

In [204]:
df


0         1         2         3         4         5         6  \
0     -13.159835 -3.055055 -2.923286 -0.238970  0.996557  0.612685 -0.422731   
1     -13.171926 -2.805815 -1.856462 -0.242773 -0.014284 -0.101964 -0.718199   
2     -13.175661 -2.546699 -0.904643 -0.221502 -0.725452 -0.051247 -0.610428   
3     -13.186564 -2.223374  0.130993 -0.306661  0.925885 -0.078233 -0.731552   
4     -13.187251 -1.940089  1.099114 -0.287128 -0.411237 -0.236615 -0.717435   
5     -13.161028 -3.106410 -2.907778 -0.172272 -0.927130  0.313955 -0.443188   
6     -13.159982 -3.119722 -2.902897 -0.129957 -1.594994  0.465704 -0.371055   
7     -13.170490 -2.523582 -0.989784 -0.153157 -0.837104  0.361975 -0.454293   
8     -13.202763 -1.644420  2.138288 -0.410963  1.973149 -0.852425 -1.120188   
9     -13.171623 -2.803745 -1.859552 -0.233100 -0.024198  0.060649 -0.648496   
10    -13.195386 -1.382345  2.918404 -0.180117 -0.566244  0.233027 -0.494365   
11    -13.195354 -1.403807  2.915277 -0.153324 -1.282392  0.325174 -0.385479   
12    -13.171841 -2.792945 -1.862761 -0.296646  0.810805 -0.249580 -0.796669   
13    -13.177870 -2.784356 -1.771148 -0.374799  1.656937 -0.647146 -1.046321   
14    -13.199968 -1.075417  3.860742 -0.275149  0.889192 -0.022763 -0.607578   
15    -13.202748 -1.655463  2.138995 -0.383442  1.466277 -0.756132 -1.051155   
16    -13.201556 -1.379676  3.007364 -0.277573  0.329234 -0.297595 -0.773314   
17    -13.193395 -2.216483  0.230489 -0.419563  1.969806 -0.935834 -1.187799   
18    -13.202784 -1.641465  2.138262 -0.419141  2.114535 -0.888117 -1.143200   
19    -13.200207 -1.116817  3.866261 -0.186364 -0.909296  0.162946 -0.428957   
20    -13.180003 -2.265344  0.046312 -0.189282 -1.053496  0.229374 -0.456751   
21    -13.160596 -3.067484 -2.914970 -0.245706  0.692828  0.261697 -0.555166   
22    -13.192596 -2.261666  0.240228 -0.311952 -0.060994 -0.678316 -0.995172   
23    -13.200230 -1.129925  3.867496 -0.155681 -1.496944  0.253070 -0.358072   
24    -13.201929 -1.680553  2.142510 -0.350407  0.611241 -0.742108 -0.997199   
25    -13.195636 -1.385770  2.912132 -0.237552 -0.096485 -0.080390 -0.593636   
26    -13.161032 -3.097431 -2.903963 -0.167399 -0.787921  0.337439 -0.470935   
27    -13.175481 -2.514039 -0.908912 -0.291970  0.696385 -0.202948 -0.753718   
28    -13.164298 -3.101386 -2.809770 -0.301965  0.158502 -0.357691 -0.506877   
29    -13.170833 -2.822773 -1.856224 -0.216181 -0.593840  0.009532 -0.638056   
...          ...       ...       ...       ...       ...       ...       ...   
45181  -4.843473  2.453041 -1.322554  2.148272  1.205305  1.232235  0.590928   
45182  -4.836158  1.881783 -3.207422  2.903635 -0.315820 -0.658298  0.001760   
45183  -4.869241  3.160505  0.765601  3.514533  3.188577 -1.525964 -0.753747   
45184  -3.858391  3.077163  0.606464  3.249049  1.837197 -0.311031 -0.244441   
45185  -3.865671  3.591134  2.421663  2.950075  1.282329  0.944729  0.284184   
45186  -3.862470  2.658601 -0.216338  1.771433  1.952223 -0.505266 -0.636418   
45187  -3.881530  3.142436  2.655667 -0.582297 -1.048928  0.387897 -0.308397   
45188  -3.843957  1.351821 -4.081177  1.331518 -0.931101 -0.135042 -0.427799   
45189  -3.871241  3.512545  2.549339  2.902877 -1.402105 -0.153759 -0.028984   
45190  -3.851050  2.359142 -1.259917  2.374367 -0.597337 -0.334360 -0.316663   
45191  -3.859215  2.918395  0.560898  1.469593  2.005202  0.888150 -0.126475   
45192  -3.837755  1.169978 -4.115948 -0.599012 -1.437604  0.438372 -0.323590   
45193  -3.864465  3.831725  3.403093  3.099030 -1.661934  0.123005  0.457868   
45194  -3.833472  1.498449 -4.224012  2.555639  1.000628 -0.609130 -0.090963   
45195  -3.859530  3.055600  0.613908  3.156047  2.169545 -0.483973 -0.282597   
45196  -3.895868  4.294889  6.476279 -0.584110 -1.745528  0.407253 -0.175674   
45197  -3.848373  1.156704 -4.026461 -0.634556 -0.437780  0.182802 -0.542616   
45198  -3.834239  1.173587 -4.126379 -0.684885 -0.592095  0.361635 -0.027237   
45199  -3.85189

In [205]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45211 entries, 0 to 45210
Data columns (total 11 columns):
0    45211 non-null float64
1    45211 non-null float64
2    45211 non-null float64
3    45211 non-null float64
4    45211 non-null float64
5    45211 non-null float64
6    45211 non-null float64
7    45211 non-null float64
8    45211 non-null float64
9    45211 non-null float64
y    45211 non-null int64
dtypes: float64(10), int64(1)
memory usage: 3.8 MB


In [206]:
df.describe()

0             1             2             3             4  \
count  4.521100e+04  4.521100e+04  4.521100e+04  4.521100e+04  4.521100e+04   
mean   8.769571e-15 -1.695645e-15  1.232382e-15 -3.272888e-16 -3.288849e-17   
std    8.365259e+00  3.173590e+00  3.104133e+00  1.229745e+00  1.087337e+00   
min   -1.322596e+01 -4.894689e+00 -6.780570e+00 -2.839595e+00 -2.603480e+00   
25%   -7.180119e+00 -2.383164e+00 -2.608297e+00 -5.730371e-01 -8.288947e-01   
50%   -9.993422e-01 -1.005309e+00 -3.946397e-02 -3.305542e-01 -1.604116e-01   
75%    7.285694e+00  1.492574e+00  2.609055e+00 -1.206442e-01  6.919513e-01   
max    1.738375e+01  1.066274e+01  8.462027e+00  8.274590e+01  1.240621e+01   

                  5             6             7             8             9  \
count  4.521100e+04  4.521100e+04  4.521100e+04  4.521100e+04  4.521100e+04   
mean   1.089117e-16  8.060910e-17 -2.363831e-16  8.081538e-16  4.918663e-18   
std    9.997448e-01  9.864658e-01  8.428928e-01  6.883861e-01  6.774870e-01   
min   -1.415291e+01 -1.601422e+00 -3.658642e+00 -2.309955e+00 -9.862233e-01   
25%   -4.414418e-01 -5.292662e-01 -5.415711e-01  7.066182e-03 -5.348553e-01   
50%   -4.290390e-02 -2.232861e-01 -2.747559e-02  1.247753e-01 -2.623658e-01   
75%    2.947590e-01  2.260162e-01  4.743622e-01  2.082446e-01  3.588609e-01   
max    2.789238e+01  1.798276e+01  8.226162e+00  8.254661e+01  2.823029e+00   

                  y  
count  45211.000000  
mean       0.116985  
std        0.321406  
min        0.000000  
25%        0.000000  
50%        0.000000  
75%        0.000000  
max        1.000000

In [207]:
#Saving as pickle file

df.to_pickle('data/bank_data_factorize_1.pkl')    # using factorization and replacing pdays only
#df.to_pickle('data/bank_data_factorize_2.pkl')    # using factorization and replacing pdays and poutcome

#df.to_pickle('data/bank_data_factorize_1_pca.pkl')    # using factorization and replacing pdays only and performing PCA
#df.to_pickle('data/bank_data_factorize_2_pca.pkl')    # using factorization and replacing pdays and poutcome and performing PCA

### Bank Additional Dataset

In [24]:
# Reading data from file

with open('bank-additional/bank-additional-full.csv', 'r') as csvfile:
    spamreader = csv.reader(csvfile, delimiter=';')
    df = list(spamreader)
    df = np.asarray(df)
    df = pd.DataFrame(data=df[1:], columns=df[0])

In [25]:
# Inspecting the unknown values in the dataset

jobs = df['job']
unknown_jobs = 0
for job in jobs:
    if job == 'unknown':
        unknown_jobs = unknown_jobs + 1
        
maritals = df['marital']
unknown_maritals = 0
for marital in maritals:
    if marital == 'unknown':
        unknown_maritals = unknown_maritals + 1

educations = df['education']
unknown_education = 0
for education in educations:
    if education == 'unknown':
        unknown_education = unknown_education + 1

defaults = df['default']
unknown_default = 0
for default in defaults:
    if default == 'unknown':
        unknown_default = unknown_default + 1

housing = df['housing']
unknown_housing = 0
for house in housing:
    if house == 'unknown':
        unknown_housing = unknown_housing + 1
    
loans = df['loan']
unknown_loans = 0
for loan in loans:
    if loan == 'unknown':
        unknown_loans = unknown_loans + 1
    
pdays = df['pdays']
unknown_pday = 0
for pday in pdays:
    if pday == '999':      # Treating never contacted as unknown
        unknown_pday = unknown_pday + 1

pouts = df['poutcome']
unknown_pouts = 0
for pout in pouts:
    if pout == 'nonexistent':
        unknown_pouts = unknown_pouts + 1
    

print("Unknown values\n")
print("Unknown jobs :", unknown_jobs)
print("Unknown marital status :", unknown_maritals)
print("Unknown education :", unknown_education)
print("Unknown default :", unknown_default)
print("Unknown housing :", unknown_housing)
print("Unknown loans :", unknown_loans)
print("Unknown pdays :", unknown_pday)
print("Unknown pouts :", unknown_pouts)


Unknown values

('Unknown jobs :', 330)
('Unknown marital status :', 80)
('Unknown education :', 1731)
('Unknown default :', 8597)
('Unknown housing :', 990)
('Unknown loans :', 990)
('Unknown pdays :', 39673)
('Unknown pouts :', 35563)


### Observations from Data Set

- Total number of rows: 41188
- Total number of attributes: 19 + output attribute (y)
- Yes samples - 4640
- No samples - 36548

- The number of yes samples are much smaller than the number of no samples indicating that the data set is highly skewed.

- Unknown values :-
    - job - 288
    - education - 1857
    - contact - 13020
    - pdays - 36954
    - poutcome - 36959
- A very large number of rows have pdays = 999 and poutcome = nonexistent. These represent clients that have not been contacted in the past

In [26]:
df = df.drop('duration', axis=1)             # Removing duration attribute

numeric = ['age', 'campaign', 'pdays', 'previous', 'emp.var.rate', 'cons.price.idx', 'cons.conf.idx', 'euribor3m', 'nr.employed']
categorical = ['job', 'marital', 'education', 'default', 'housing', 'loan', 'contact', 'month', 'day_of_week', 'poutcome']


In [27]:
# Taking care of categorical values

# factorize assigns unique numeric values to each category
for c in categorical:
    df[c] = pd.factorize(df[c])[0]
    
# Use dummy attributes for categorical values
# df = pd.get_dummies(df, columns=categorical, prefix=categorical)
# df['y'] = pd.factorize(df['y'])

df[numeric] = df[numeric].astype('float')
df[categorical] = df[categorical].astype('category')

# df['y'] = df['y'].astype(int) # Converting otuput variable to integers

In [28]:
# Normalizing numerical values

scaler = preprocessing.StandardScaler()
rescaled_df = scaler.fit_transform(df[numeric])
df[numeric] = pd.DataFrame(rescaled_df, columns=numeric)

In [29]:
# Principal Component Analysis

y_values = df.loc[:, ['y']].values
df.drop(columns=['y'], axis=1, inplace=True)

#pca = PCA()
pca = PCA(n_components=10)
components = pca.fit_transform(df)
var_ratios = pca.explained_variance_ratio_
#rint(var_ratios)

# Calculating the number of attributes for 80% variance
s=0
count = 0
for ratio in var_ratios:
    s = s + ratio
    count = count + 1
    if s >= 0.8:
        break
print(count)

# Calculating the number of attributes for 95% variance
s=0
count1 = 0
for ratio in var_ratios:
    s = s + ratio
    count1 = count1+ 1
    if s >= 0.95:
        break
print(count1)

df = pd.DataFrame(data=components)
df['y'] = y_values

6
10


### Observations from PCA

- There were 18 attributes (after removing duration) subjected to PCA
- 80% of the variance can be captured by 6 attributes after PCA
- 95% of the variance can be captured by 10 attributes after PCA


In [30]:
df

0         1         2         3         4         5         6  \
0     -5.085238  1.976573  1.192552 -0.215155 -2.003972  0.848267  1.607368   
1     -4.245930  0.958810  0.738160 -0.208968 -2.013441  0.868691  1.707951   
2     -4.239274  0.939066  0.595273 -0.055896 -1.964206  0.416226  0.453556   
3     -3.379089 -0.068760  0.134065 -0.036606 -1.974160  0.474288  0.637311   
4     -4.220810  0.969717  0.712080 -0.169690 -2.008423  0.826473  1.603210   
5     -3.569816  0.414489 -1.113143  0.100325 -2.003916  0.617640  1.020843   
6     -2.653864 -0.550303 -1.523710  0.097162 -2.043962  0.868999  1.832070   
7     -1.844656 -1.606542 -2.096716  0.210928 -2.004811  0.493633  0.779250   
8     -1.651643 -2.102220 -0.944987  0.173770 -1.949430  0.052176 -0.447901   
9     -4.247304  0.919201  0.499158  0.036175 -1.935292  0.134031 -0.389039   
10    -1.844656 -1.606542 -2.096716  0.210928 -2.004811  0.493633  0.779250   
11    -4.247304  0.919201  0.499158  0.036175 -1.935292  0.134031 -0.389039   
12    -3.216231 -0.573435  1.340109 -0.146080 -1.931074  0.192210 -0.190093   
13    -5.051376  1.994906  1.157028 -0.164862 -2.013406  0.788744  1.438788   
14    -2.858508 -0.817904  0.512381 -0.077291 -1.959628  0.334420  0.312696   
15    -1.473820 -2.546060  0.578083 -0.248230 -2.009017  0.708548  1.481140   
16    -2.858508 -0.817904  0.512381 -0.077291 -1.959628  0.334420  0.312696   
17    -2.870428 -0.809341  0.598547 -0.170215 -1.988006  0.582043  1.014946   
18    -2.488504 -1.047929 -0.283343 -0.040060 -2.007489  0.644894  1.239183   
19    -0.977410 -3.325501  0.872668 -0.225058 -1.965228  0.361422  0.442162   
20    -1.151562 -3.561276  3.004300 -0.484141 -1.912852  0.167507 -0.090739   
21    -3.549973 -0.270704  2.432293 -0.464008 -1.990968  0.772053  1.524710   
22    -2.127032 -2.005866  2.322019 -0.440472 -1.991958  0.679755  1.351867   
23    -2.667398 -1.287763  1.825725 -0.284507 -1.961158  0.407922  0.522056   
24    -3.721088  0.192611  1.004167 -0.132592 -1.959180  0.396424  0.433492   
25    -0.961709 -2.612227 -2.633312  0.322494 -1.992564  0.326919  0.371598   
26    -1.259614 -2.295788 -1.593481  0.072145 -2.047790  0.812945  1.811596   
27     0.047981 -4.822847  1.711657 -0.402576 -1.953383  0.354040  0.516165   
28    -0.962356 -2.593898 -2.524118  0.179458 -2.037801  0.718942  1.350054   
29     1.652190 -6.302841 -0.445218 -0.189466 -2.025421  0.631190  1.517829   
...         ...       ...       ...       ...       ...       ...       ...   
41158  1.455823  2.563635  3.583121 -0.068598 -0.935472 -1.121736 -1.888441   
41159  2.959442  3.559434  4.334583  3.654009 -0.994653  6.359505 -4.925692   
41160  2.479693  2.483561 -2.533917  0.771909 -1.000365 -1.062586 -1.617394   
41161  2.771732  2.688360 -2.384066  1.489784 -1.003474  0.100078 -2.177415   
41162  1.762411  3.860179  3.752429  1.796024 -1.054733  4.120591 -1.878188   
41163  2.475882  3.219384  4.076369  2.458061 -0.981015  4.478407 -4.212797   
41164  2.611781  3.545337 -0.248605  2.374749 -1.083132  4.009594 -2.132616   
41165  1.464973  3.989302 -3.381138  0.912771 -0.020011 -0.958884 -1.583936   
41166  2.691342  2.621573 -2.404569  1.441859  0.010129  0.192921 -2.194344   
41167  2.465168  2.471153 -2.565826  0.726832  0.013328 -1.067874 -1.865909   
41168  6.624872 -3.484121  0.750203  0.097556  0.033831 -1.092461 -1.556311   
41169  0.226649  4.555107  1.637425 -0.021638 -0.000240 -0.406308 -0.337777   
41170  5.553155  0.196624 -0.400704  2.883834  0.003928  2.993940 -3.534326   
41171  6.396257 -3.759371  2.899010 -0.205852  1.069641 -1.139968 -1.774780   
41172  2.510802  2.501902 -2.587674  0.800591  0.997949 -1.117158 -1.808228   
41173  4.613166  0.677310 -0.875469  1.655922  0.936343  1.635765 -0.895218   
41174  6.056529  1.629604 -0.150591  5.244802  0.880475  8.936201 -3.844252   
41175  7.341344 -3.593677  2.270087  1.257735  1.046944  0.902002 -2.850159   
41176 -0.242632  5.313085  1.056839  0.336594  1.022262 -0.96271

In [21]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 41188 entries, 0 to 41187
Data columns (total 20 columns):
age               41188 non-null float64
job               41188 non-null category
marital           41188 non-null category
education         41188 non-null category
default           41188 non-null category
housing           41188 non-null category
loan              41188 non-null category
contact           41188 non-null category
month             41188 non-null category
day_of_week       41188 non-null category
campaign          41188 non-null float64
pdays             41188 non-null float64
previous          41188 non-null float64
poutcome          41188 non-null category
emp.var.rate      41188 non-null float64
cons.price.idx    41188 non-null float64
cons.conf.idx     41188 non-null float64
euribor3m         41188 non-null float64
nr.employed       41188 non-null float64
y                 41188 non-null object
dtypes: category(10), float64(9), object(1)
memory usage: 3.5+

In [22]:
df.describe()

age      campaign         pdays      previous  emp.var.rate  \
count  4.118800e+04  4.118800e+04  4.118800e+04  4.118800e+04  4.118800e+04   
mean   2.846557e-16  1.845337e-14  9.515119e-14  2.196991e-13 -1.851020e-13   
std    1.000012e+00  1.000012e+00  1.000012e+00  1.000012e+00  1.000012e+00   
min   -2.209365e+00 -5.659220e-01 -5.149443e+00 -3.494943e-01 -2.216433e+00   
25%   -7.699804e-01 -5.659220e-01  1.954139e-01 -3.494943e-01 -1.197935e+00   
50%   -1.942267e-01 -2.049085e-01  1.954139e-01 -3.494943e-01  6.480923e-01   
75%    6.694038e-01  1.561049e-01  1.954139e-01 -3.494943e-01  8.390606e-01   
max    5.563310e+00  1.928982e+01  1.954139e-01  1.379492e+01  8.390606e-01   

       cons.price.idx  cons.conf.idx     euribor3m   nr.employed  
count    4.118800e+04   4.118800e+04  4.118800e+04  4.118800e+04  
mean    -6.434047e-14  -6.842191e-14 -6.864230e-14 -1.181213e-13  
std      1.000012e+00   1.000012e+00  1.000012e+00  1.000012e+00  
min     -2.374889e+00  -2.224953e+00 -1.722351e+00 -2.815697e+00  
25%     -8.649547e-01  -4.747909e-01 -1.312994e+00 -9.402810e-01  
50%      2.994571e-01  -2.803284e-01  7.124599e-01  3.316799e-01  
75%      7.227225e-01   8.864466e-01  7.724221e-01  8.451701e-01  
max      2.058168e+00   2.939106e+00  8.208531e-01  8.451701e-01

In [31]:
#Saving as pickle file
df.to_pickle('data/bank_additional_data_factorize_pca.pkl')